src/validator.ts

```diff
@@ -2,49 +2,243 @@ import Ajv from 'ajv';
 import type { ErrorObject } from 'ajv';
 import { PineconeArgumentError } from './errors';
 
-export const buildValidator = (
-  schema: any,
-  onError: (errorsList: string[]) => void
+const prepend = (prefix: string, message: string) => {
+  return `${prefix} ${message}`;
+};
+
+const schemaPathPropNameRegex = /properties\/(.+)\//;
+const schemaPathArrayPropNameRegex = /properties\/(.+)\/items/;
+const schemaPathGroupNumberRegex = /anyOf\/(\d+)\/(.+)/;
+const instancePathItemIndexRegex = /(\d+)$/;
+
+// If there are more than maxErrors errors in a group, they
+// will get summarized with an error count.
+const maxErrors = 3;
+
+const formatIndividualError = (e, formattedMessageList) => {
+  if (e.schemaPath.indexOf('properties') > -1) {
+    // property of an object
+    if (e.schemaPath.indexOf('items') > -1) {
+      // property is an array
+      const propNameMatch = schemaPathArrayPropNameRegex.exec(e.schemaPath);
+      const propName = propNameMatch ? propNameMatch[1] : 'unknown';
+      const itemIndexMatch = instancePathItemIndexRegex.exec(e.instancePath);
+      const itemIndex = itemIndexMatch ? itemIndexMatch[1] : 'unknown';
+      formattedMessageList.push(
+        `item at index ${itemIndex} of the '${propName}' array ${e.message}`
+      );
+    } else {
+      // property is not an array
+      const propNameMatch = schemaPathPropNameRegex.exec(e.schemaPath);
+      const propName = propNameMatch ? propNameMatch[1] : 'unknown';
+      formattedMessageList.push(`property '${propName}' ${e.message}`);
+    }
+  } else if (e.schemaPath.indexOf('items') > -1) {
+    // item in an array
+    const itemIndexMatch = instancePathItemIndexRegex.exec(e.instancePath);
+    const itemIndex = itemIndexMatch ? itemIndexMatch[1] : 'unknown';
+    formattedMessageList.push(
+      `item at index ${itemIndex} of the array ${e.message}`
+    );
+  } else if (e.instancePath === '') {
+    // parameter is something other than an object, e.g. string
+    formattedMessageList.push(`the argument ${e.message}`);
```

Comment:
**micro-nit** This is _such_ a particular nit, but I kind of think this reads a bit better.

```suggestion
    formattedMessageList.push(`argument ${e.message}`);
```